crates/ruff/src/rules/ruff/rules/implicit_optional.rs

```diff
@@ -180,60 +179,75 @@ impl<'a> TypingTarget<'a> {
+                // In case of a parse error, we return `None` to avoid false positives.
+                .map_or(Some(TypingTarget::None), |(expr, _)| {
+                    Some(TypingTarget::ForwardReference(expr))
+                }),
+            _ => checker
+                .semantic()
+                .resolve_call_path(expr)
+                .and_then(|call_path| {
+                    if checker.semantic().match_typing_call_path(&call_path, "Any") {
+                        Some(TypingTarget::Any)
+                    } else if matches!(call_path.as_slice(), ["" | "builtins", "object"]) {
+                        Some(TypingTarget::Object)
+                    } else {
+                        None
+                    }
+                }),
         }
     }
 
     /// Check if the [`TypingTarget`] explicitly allows `None`.
-    fn contains_none(&self, semantic: &SemanticModel) -> bool {
+    fn contains_none(&self, checker: &Checker) -> bool {
         match self {
             TypingTarget::None
             | TypingTarget::Optional
             | TypingTarget::Any
             | TypingTarget::Object => true,
             TypingTarget::Literal(elements) => elements.iter().any(|element| {
-                let Some(new_target) = TypingTarget::try_from_expr(element, semantic) else {
+                let Some(new_target) = TypingTarget::try_from_expr(element, checker) else {
                     return false;
                 };
                 // Literal can only contain `None`, a literal value, other `Literal`
                 // or an enum value.
                 match new_target {
                     TypingTarget::None => true,
-                    TypingTarget::Literal(_) => new_target.contains_none(semantic),
+                    TypingTarget::Literal(_) => new_target.contains_none(checker),
                     _ => false,
                 }
             }),
             TypingTarget::Union(elements) => elements.iter().any(|element| {
-                let Some(new_target) = TypingTarget::try_from_expr(element, semantic) else {
+                let Some(new_target) = TypingTarget::try_from_expr(element, checker) else {
                     return false;
                 };
                 match new_target {
                     TypingTarget::None => true,
-                    _ => new_target.contains_none(semantic),
+                    _ => new_target.contains_none(checker),
                 }
             }),
             TypingTarget::Annotated(element) => {
-                let Some(new_target) = TypingTarget::try_from_expr(element, semantic) else {
+                let Some(new_target) = TypingTarget::try_from_expr(element, checker) else {
+                    return false;
+                };
+                match new_target {
+                    TypingTarget::None => true,
+                    _ => new_target.contains_none(checker),
+                }
+            }
+            TypingTarget::ForwardReference(expr) => {
```

Comment:
Would it be possible to also take `&Expr` in `ForwardReference` and unify the branches or doesn't that work because the expression is created outside the AST an needs to be owned?

Reply:
I was unable to unify them due to the owned value. For `ForwardReference`, the enum needs to own it while for others we receive it via the AST checker.

Reply:
Oh, I found one small refactor :)